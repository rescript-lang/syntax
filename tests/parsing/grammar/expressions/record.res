let r = {a: expr}
let r = {a: expr,} // trailing comma

let r = {Parsetree.pexp_attributes: [], Parsetree.loc: loc}

// punning
let r = {a, b, c} 
let r = {A.a, b}
let r = {A.a, b, C.c}

let r = {Parsetree.pexp_attributes, Parsetree.loc}
// trailing comma
let r = {Parsetree.pexp_attributes, Parsetree.loc,}

// with constraint
let r = {a: (expr : int), b: (x : string)}

// spread
let r = {...expr, pexp_attributes: []}
let r = {...expr, pexp_attributes: [], pexp_loc: loc}
let r = {...expr, pexp_attributes: [],} // trailing comma

// with type constraint on spread
let r = {...make() : myRecord, foo: bar}
let r = {...(make() : myRecord), foo: bar} // parens optional

let r = {x: @optional None, y: @optional None, z: @optional (None:tt)}

let z = name => { name : @optional name, x: 3}

let z = name => { @optional name, x: 3}

let z = name => { name, @optional x }

let _ = switch z {
  | {x: @optional None, y: @optional None, z: @optional (None:tt)} => 11
  | {name:  @optional name, x: 3} => 42
  | {@optional name, x: 3} => 4242
}