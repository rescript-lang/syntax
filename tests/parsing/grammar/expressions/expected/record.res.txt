let r = { a = expr }
let r = { a = expr }
let r = { Parsetree.pexp_attributes = [||]; Parsetree.loc = loc }
let r = { a; b; c }
let r = { A.a = a; b }
let r = { A.a = a; b; C.c = c }
let r = { Parsetree.pexp_attributes = pexp_attributes; Parsetree.loc = loc }
let r = { Parsetree.pexp_attributes = pexp_attributes; Parsetree.loc = loc }
let r = { a = (expr : int); b = (x : string) }
let r = { expr with pexp_attributes = [||] }
let r = { expr with pexp_attributes = [||]; pexp_loc = loc }
let r = { expr with pexp_attributes = [||] }
let r = { (make () : myRecord) with foo = bar }
let r = { (make () : myRecord) with foo = bar }
let r =
  {
    x = ((None)[@optional ]);
    y = ((None)[@optional ]);
    z = (((None : tt))[@optional ])
  }
let z name = { name = ((name)[@optional ]); x = 3 }
let z name = { name = ((name)[@optional ]); x = 3 }
let z name = { name; x = ((x)[@optional ]) }
let _ =
  match z with
  | { x = ((None)[@optional ]); y = ((None)[@optional ]);
      z = (((None : tt))[@optional ]) } -> 11
  | { name = ((name)[@optional ]); x = 3 } -> 42
  | { name = ((name)[@optional ]); x = 3 } -> 4242